crates/compiler/can/src/desugar.rs

```diff
@@ -137,230 +181,209 @@ pub fn desugar_defs_node_values<'a>(
-                                    )),
-                                )
-                            } else {
-                                loc_ret
-                            }
-                        };
-
-                        // Unwrap Suffixed def within Apply, and the pattern so we can use in the call to Task.await
-                        let (suffixed_sub_apply_loc, pattern) = unwrap_suffixed_def_and_pattern(
-                            arena,
-                            loc_expr.region,
-                            defs.value_defs[value_index],
-                        );
-
-                        // Create Closure for the result of the recursion,
-                        // use the pattern from our Suffixed Def as closure argument
-                        let closure_expr =
-                            Closure(arena.alloc([*pattern]), defs_after_suffixed_desugared);
-
-                        // Apply arguments to Task.await, first is the unwrapped Suffix expr second is the Closure
-                        let mut task_await_apply_args: Vec<&'a Loc<Expr<'a>>> = Vec::new_in(arena);
-
-                        task_await_apply_args
-                            .push(arena.alloc(Loc::at(loc_expr.region, suffixed_sub_apply_loc)));
-                        task_await_apply_args
-                            .push(arena.alloc(Loc::at(loc_expr.region, closure_expr)));
-
-                        let new_loc_return = arena.alloc(Loc::at(
-                            loc_expr.region,
-                            Apply(
-                                arena.alloc(Loc {
-                                    region: loc_expr.region,
-                                    value: Var {
-                                        module_name: ModuleName::TASK,
-                                        ident: "await",
-                                    },
-                                }),
-                                arena.alloc(task_await_apply_args),
-                                CalledVia::BangSuffix,
-                            ),
-                        ));
-
-                        arena.alloc(Loc::at(
-                            loc_expr.region,
-                            Defs(arena.alloc(before), new_loc_return),
-                        ))
-                    }
-                }
-            }
+        // we only want to unwrap suffixed nodes if this is a top level def
+        // this function will be called recursively in desugar_expr and we only
+        // unwrap expression after they have completed desugaring to simplify the analysis
+        if top_level_def {
+            // first desugar, then unwrap any suffixed defs
+            *value_def = desugar_defs_node_values_suffixed(
+                arena,
+                desugar_value_def(arena, arena.alloc(*value_def), src, line_info, module_path),
+                src,
+                line_info,
+                module_path,
+            );
+        } else {
```

Comment:
This is where we being unwrapping expression for top_level_defs only. It's a single pass through the AST after it has been desugared (removed spaces etc). It descends through the AST and each time it finds a suffixed expression it will unwrap that, and then start from the top again, so unfortunately this grows asymptotically.